app/script/entity/message/File.js

```diff
@@ -0,0 +1,176 @@
+/*
+ * Wire
+ * Copyright (C) 2017 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.entity = z.entity || {};
+
+z.entity.File = class File extends z.entity.Asset {
+  constructor(id) {
+    super(id);
+    this.type = z.assets.AssetType.FILE;
+    this.logger = new z.util.Logger('z.entity.File', z.config.LOGGER.OPTIONS);
+
+    // z.assets.AssetTransferState
+    this.status = ko.observable();
+
+    this.file_name = '';
+    this.file_size = '';
+    this.file_type = '';
+
+    // contains asset meta data as object
+    this.meta = {};
+
+    // asset url, instance of an otr asset this has to be decrypted
+    this.original_resource = ko.observable();
+    this.preview_resource = ko.observable();
+
+    this.download_progress = ko.pureComputed(() => {
+      if (this.original_resource()) {
+        return this.original_resource().download_progress();
+      }
+
+      return undefined;
+    });
+
+    this.upload_id = ko.observable();
+    this.upload_progress = ko.observable();
+    this.uploaded_on_this_client = ko.observable(false);
+    this.upload_failed_reason = ko.observable();
+    this.upload_cancel = undefined;
+    this.pending_upload = ko.pureComputed(() => {
+      return (this.status() === z.assets.AssetTransferState.UPLOADING) && this.uploaded_on_this_client();
+    });
+
+    // update progress
+    this.upload_id.subscribe((upload_id) => {
+      if (upload_id) {
+        return amplify.subscribe(`upload${upload_id}`, this.on_progress);
+      }
+    });
+
+    this.status.subscribe((status) => {
+      if (status === z.assets.AssetTransferState.UPLOADED) {
+        return amplify.unsubscribe(`upload${this.upload_id}`, this.on_progress);
+      }
+    });
+  }
+
+  on_progress(progress) {
+    return this.upload_progress(progress);
+  }
+
+  /**
+   * Loads and decrypts otr asset preview
+   *
+   * @returns {Promise} Returns a promise that resolves with the asset as blob
+   */
+  load_preview() {
+    if (this.preview_resource()) {
+      return this.preview_resource().load();
+    }
+
+    return undefined;
```

Comment:
It if is supposed to return a promise we should not return undefined but Promise.resolve or Promise.reject. Maybe a return value is not needed though.